File: Number/1. Two Sum.js
/*

這段程式碼的時間複雜度是 O(n)，其中 n 是 nums 的長度。

*/

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (nums, target) {

  map = {};
  for (i = 0; i < nums.length; i++) {
      var v = nums[i];
      if (map[target - v] >= 0) {
          return [map[target - v], i];
      } else {
          map[v] = i;
      }
  }
};

File: Array/1. Two Sum.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        nums_dic = {}
        for index,num in enumerate(nums):
            complement = target - num

            if complement in nums_dic:
                return [nums_dic[complement],index]
            
            nums_dic[num] = index